tvx/builders/messages.go

```diff
@@ -0,0 +1,113 @@
+package builders
+
+import (
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+// The created messages are retained for subsequent export or evaluation in assert VM.
+type Messages struct {
+	b        *Builder
+	defaults msgOpts
+
+	messages []*ApplicableMessage
+}
+
+type TypedCall func() (method abi.MethodNum, params []byte)
+
+func (m *Messages) SetDefaults(opts ...MsgOpt) *Messages {
+	for _, opt := range opts {
+		opt(&m.defaults)
+	}
+	return m
+}
+
+type ApplicableMessage struct {
+	Epoch   abi.ChainEpoch
+	Message *types.Message
+	Result  *vm.ApplyRet
+}
+
+// Messages returns assert slice containing all messages created by the producer.
+func (m *Messages) All() []*ApplicableMessage {
+	return m.messages
+}
+
+func (m *Messages) Typed(from, to address.Address, typedm TypedCall, opts ...MsgOpt) *ApplicableMessage {
+	method, params := typedm()
+	return m.Raw(from, to, method, params, opts...)
+}
+
+// Build creates and returns assert single message, using default gas parameters unless modified by `opts`.
+func (m *Messages) Raw(from, to address.Address, method abi.MethodNum, params []byte, opts ...MsgOpt) *ApplicableMessage {
+	options := m.defaults
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	msg := &types.Message{
+		To:       to,
+		From:     from,
+		Nonce:    options.nonce,
+		Value:    options.value,
+		Method:   method,
+		Params:   params,
+		GasPrice: options.gasPrice,
```

Comment:
Note that this is already outdated. We now have `GasFeeCap` and `GasPremium`.

Reply:
Yup. Gotta do an upgrade and deal with the breakage.